control-plane/piSerializer.cpp

```diff
@@ -0,0 +1,1471 @@
+  auto tableSize = getTableSize(table);
+  auto defaultAction = getDefaultAction(table, refMap, typeMap);
+  auto matchFields = getMatchFields(table, refMap, typeMap);
+
+  std::vector<cstring> actions;
+  for (auto action : *table->getActionList()->actionList) {
+    auto decl = refMap->getDeclaration(action->getPath(), true);
+    BUG_CHECK(decl->is<IR::P4Action>(), "Not an action: '%1%'", decl);
+    actions.push_back(decl->to<IR::P4Action>()->externalName());
+  }
+
+  serializer.addTable(table, tableSize, defaultAction, actions, matchFields);
+}
+
+/// Visit evaluated blocks under the provided top-level block. Guarantees that
+/// each block is visited only once, even if multiple paths to reach it exist.
+template <typename Func>
+static void forAllEvaluatedBlocks(const IR::ToplevelBlock* aToplevelBlock,
+                                  Func function) {
+  set<const IR::Block*> visited;
+  ordered_set<const IR::Block*> frontier{aToplevelBlock};
+
+  while (!frontier.empty()) {
+    // Pop a block off the frontier of blocks we haven't yet visited.
+    auto evaluatedBlock = *frontier.begin();
+    frontier.erase(frontier.begin());
+    visited.insert(evaluatedBlock);
+
+    function(evaluatedBlock);
+
+    // Add child blocks to the frontier if we haven't already visited them.
+    for (auto evaluatedChild : evaluatedBlock->constantValue) {
+      if (!evaluatedChild.second->is<IR::Block>()) continue;
+      auto evaluatedChildBlock = evaluatedChild.second->to<IR::Block>();
+      if (visited.find(evaluatedChildBlock) != visited.end()) continue;
+      frontier.insert(evaluatedChildBlock);
+    }
+  }
+}
+
+} // namespace ControlPlaneAPI
+
+void serializeToPI(std::ostream* destination,
+                   const IR::P4Program* program,
+                   const IR::ToplevelBlock* evaluatedProgram,
+                   ReferenceMap* refMap,
+                   TypeMap* typeMap,
+                   PIFormat format /* = PIFormat::BINARY */) {
+  using namespace ControlPlaneAPI;
+
+  PISymbolTable symbols = PISymbolTable::create([=](PISymbolTable& symbols) {
+    // Synthesize a ".drop" symbol.
+    symbols.add(PISymbolType::ACTION, ".drop");
+
+    // Perform a first pass to collect all of the symbols in the program.
+    forAllEvaluatedBlocks(evaluatedProgram, [&](const IR::Block* block) {
+      if (block->is<IR::ControlBlock>()) {
+        collectControlSymbols(symbols, block->to<IR::ControlBlock>(), refMap, typeMap);
+      } else if (block->is<IR::ExternBlock>()) {
+        collectExternSymbols(symbols, block->to<IR::ExternBlock>());
+      } else if (block->is<IR::ParserBlock>()) {
+        collectParserSymbols(symbols, block->to<IR::ParserBlock>(), typeMap);
+      } else if (block->is<IR::TableBlock>()) {
+        collectTableSymbols(symbols, block->to<IR::TableBlock>());
+      }
+    });
+  });
+
+  P4InfoSerializer serializer(symbols, refMap, typeMap);
+
+  // Synthesize a ".drop" action.
+  serializer.addSynthesizedAction(".drop");
```

Comment:
Totally agreed. That's another thing I intended to do immediately after this PR landed.